sway-core/src/semantic_analysis/ast_node/expression/typed_expression.rs

```diff
@@ -367,8 +390,18 @@ impl TypedExpression {
                 }
                 buf
             }
+            TypedExpressionVariant::Reassignment(reassignment) => {
+                reassignment.rhs.gather_return_statements()
+            }
+            TypedExpressionVariant::StorageReassignment(_storage_reassignment) => {
+                // FIXME: we should check rhs here too, no?
+                vec![]
+            }
             // if it is impossible for an expression to contain a return _statement_ (not an
             // implicit return!), put it in the pattern below.
+
+            // FIXME: lots of these expression kinds can contain arbitrary sub-expressions. But
+            // we're not checking those for returns?
```

Comment:
We also don't check any of the expression kinds below for return statements, even though many of them could contain return statements in sub-expressions.

Reply:
I've modified this to also recurse into all the sub-expression kinds that weren't being recursed into.